### PR TITLE
ROX-20012: Add polling for administration events

### PR DESCRIPTION
**Pull request by pedrottimark**
### Objective

Let users decide when to load new events, following the precedent of Network Graph.

Events page implicitly reloads after any user action:
* to change search filters, sorting, or (theoretically in this case) pagination
* to visit an event page and then go back

If all events in polling request are new, then render **Load 10+ new events** (or whatever is the `perPage` amount) because frontend knowledge is limited to current page.

As discussed, frontend knowledge is limited to current request params.

### Changes

1. Edit AdministrationEventsPage.tsx file.
    * Separate `useEffect` for initial request and user-initiated actions from `useEffect` for polling requests.
    * Frontend polling interval 60 seconds corresponds to backend reprocessing of events. 
    * Count available events compared to set of event id for the current page.

2. Edit AdministrationEventsToolbar.tsx file.
    * Render `UpdatedTimeOrUpdateButton` element adjacent to pagination.
    * Change from full to compact pagination according to design.

3. Add UpdatedTimeOrUpdateButton.tsx file.
    * Render either `lastUpdatedTime` paragraph or `countAvailable` button.
    * Specify 14px small size for paragraph and button same for visual compatibility with pagination.

4. Edit AdministationEventsService.ts file.
    * Add `getListAdministrationEventsArg` function to encapsulate comversion from URL hook values.

5. Edit dateUtils.ts file.
    * Add `getTimeHoursMinutes` function.

### Residue

1. Call `getTimeHoursMinutes` function:
    * https://github.com/stackrox/stackrox/blob/master/ui/apps/platform/src/Containers/NetworkGraph/NetworkGraphPage.tsx#L221-L227
    * https://github.com/stackrox/stackrox/blob/master/ui/apps/platform/src/Containers/Dashboard/SummaryCounts.tsx#L59
2. Ask at weekly joint team meeting about pro and con of date in **Last updated** on main dashboard.

## Checklist
- [x] Investigated and inspected CI test results
- [x] ~Unit test and regression tests added~

## Testing Performed

Before `yarn deploy` in ui, do the following:

```sh
export ROX_ADMINISTRATION_EVENTS=true
```

1. `yarn lint` in ui
2. `yarn build` in ui and then compute branch - master for the following
    * `wc build/static/js/*.js`
        main.js: 43 = 3893514 - 3893471
        total: 8871 = 10165134 - 10156263
    * `ls -al build/static/js/*.js | wc -l`
        1 = 83 - 82 files
3. `yarn start` in ui

### Manual testing

Pictures at 1280px width and 800px height like an older laptop.

1. Visit /main/administration-events

    * See count and list request.
    * Without `if (pollingCount !== 0) { … }` see a redundant polling request at initial page visit.
    * See `lastUpdatedTime` at the left of compact pagination.
        Visit

    * With temporary `let nAvailable = pollingCount` see update button with increasing `countAvailable` after each polling interval.
        countAvailable

    * See **10+ events available** when all events in polling request are new (with temporary hack).
        isAvailableEqualToPerPage

    * See `lastUpdatedTime` replace `countAvailable` button after click update button.
        click

    * Ditto after any user action to change search filters, sorting, or (theoretically in this case) pagination, also to visit an event page and then go back.
        filter

### Integration testing

Before `yarn cypress-open` in ui/apps/platform, do the following:

```sh
export CYPRESS_ROX_ADMINISTRATION_EVENTS=true
```

* administation/events/eventsTable.test.js

### CODE REVIEW

**Comment by pedrottimark** [comment]
gke-ui-e2e-tests failures

> Policy wizard, Step 3 Policy Criteria Existing values should populate boolean radio buttons

Screenshot looks like MainPage waiting for prerequisite requests and possibly relevant from central.log file:

pkg/search/postgres: 2023/10/10 17:49:05.914781 common.go:825: Error: error committing cursor transaction: conn closed
root logger: 2023/10/10 17:49:08.594407 logger.go:77: Warn: pkg/grpc/authn/interceptor.go:30 - Cannot extract identity: failed to read roles: loading role "Admin": timeout: context canceled

> Workload CVE watched images flow should allow user to unwatch an image via the images table row action

Fixed in #8051

File: ui/apps/platform/src/Containers/Administration/Events/AdministrationEventsPage.tsx [hunk 1]
```diff
@@ -23,41 +24,83 @@ function AdministrationEventsPage(): ReactElement {
...
-            .then((countArg) => {
-                setCount(countArg);
-            })
-            .catch(() => {
-                setCount(0);
-            });
-    }, [page, perPage, searchFilter, sortOption, setIsLoading]);
+    function updateEvents() {
+        setUpdatedCount(updatedCount + 1);
+    }
```
[the remark this answers is not in the document]
Comment (pedrottimark) [reply ui/apps/platform/src/Containers/Administration/Events/AdministrationEventsPage.tsx:74]: Good question. That was my original thought, so having this to compare, I will rethink it again that way.

File: ui/apps/platform/src/Containers/Administration/Events/AdministrationEventsPage.tsx [hunk 2]
```diff
@@ -23,41 +24,83 @@ function AdministrationEventsPage(): ReactElement {
...
+                })
+                .catch(() => {});
+        }
+    }, [pollingCount]); // eslint-disable-line react-hooks/exhaustive-deps
+    // Why disable the rule for the following hook dependencies: lastUpdatedEventIds, page, perPage, searchFilter, sortOption?
+    // So polling continues on its schedule instead of immediately making a redundant request.
+
+    useInterval(() => {
+        setPollingCount(pollingCount + 1);
+    }, 60000); // 60 seconds corresponds to backend reprocessing events.
```
[the remark this answers is not in the document]
Comment (pedrottimark) [reply ui/apps/platform/src/Containers/Administration/Events/AdministrationEventsPage.tsx:102]: Wow, that sounds super!

File: ui/apps/platform/src/Containers/Administration/Events/AdministrationEventsPage.tsx [hunk 3]
```diff
@@ -23,41 +24,83 @@ function AdministrationEventsPage(): ReactElement {
...
+    /*
+     * Request count and events at initial page load or after user action:
+     * change pagination, searchFilter, sortOptions
+     * click events available button
+     */
     useEffect(() => {
-        setIsLoading(true);
+        if (updatedCount === 0) {
+            setIsLoading(true);
+        }
```
[the remark this answers is not in the document]
Comment (pedrottimark) [reply ui/apps/platform/src/Containers/Administration/Events/AdministrationEventsPage.tsx:46]: Aha, that would fill a gap with `isDisabled={isLoading} isLoading={isLoading}` props for the **events available** button.

File: ui/apps/platform/src/Containers/Administration/Events/AdministrationEventsPage.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (vjwilson) [reply ui/apps/platform/src/Containers/Administration/Events/AdministrationEventsPage.tsx:74]: IIRC, we use a separate function called by 2 different useEffect hooks on the Clusters table page.

File: ui/apps/platform/src/services/AdministrationEventsService.ts [hunk 4]
```diff
@@ -139,6 +139,24 @@ export function listAdministrationEvents(
         .then((response) => response.data.events);
 }
 
+type GetAdministrationEventResponseArg = {
```
Comment (vjwilson) [inline ui/apps/platform/src/services/AdministrationEventsService.ts:142]: nit: putting "Response" at the end of the type name would emphasize that it is the return value of that function

```suggestion
type GetAdministrationEventArgResponse = {
```

File: ui/apps/platform/src/utils/dateUtils.ts [hunk 5]
```diff
@@ -32,6 +32,15 @@ export function getTime(timestamp: DateLike) {
...
 }
 
+/**
+ * Returns a formatted time with hours and minutes but without seconds.
+ * @param {DateLike} timestamp - The timestamp for the date
+ * @returns {string} - returns a formatted string for the time
+ */
+export function getTimeHoursMinutes(timestamp: DateLike) {
+    return format(timestamp, 'h:mmA');
+}
```
Comment (vjwilson) [inline ui/apps/platform/src/utils/dateUtils.ts:42]: nit: unit test would be good in a follow-up

**Review by vjwilson — APPROVED** [review]

**Review by dvail — APPROVED** [review]

**Pull request merged**